rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+
+### Code structure
+The code structure will be standardised. As well as using the [default Rails directory structure], we propose the following standard for additional classes (e.g. services, api calls). The standard expands on the [conventions] documented in the developer docs.
+
+The host app will follow the default MVC structure of Rails applications, with some additional restrictions on what goes where that are frontend specific (listed below)
+
+```
+├── app
+│   ├── assets
+│   ├── controllers
+│   ├── helpers
+│   ├── models
+│   ├── presenters
+│   ├── services
+│   ├── views
+├── config
+│   ├── data
+│   ├── locales
+├── lib
+│   ├── tasks
+└── .gitignore
+```
+
+Under `/app`:
+
+`models`
+- Frontend apps do not have database records, so models here are created to store data received from api calls (eg content items from content store), process data that will be written to apis (eg feedback items to zendesk), and information ingested from YAML/JSON/CSVs.
+
+`services`
+- Api calls for search and content-store, and apis external to GOV.UK
+
+`presenters`
+- Presentation related transformations on model data. It is a wrapper around a model for use in a view.
+
+`helpers`
+- General utility methods for a view.
+
+Under `/lib`
+
+`tasks`
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
+
+`locales`
+- Static content like lists of contact information should ideally be stored in the locale files to promote localisation.
+
+Namespaces will be used to group functionality e.g. for StepBySteps, and resources in the routes file. This structure will be documented in README of the host app.
+Each application will be analysed before being consolidated to ensure redundant code is not being moved over.
+
+### How the proposal will be carried out
+This is obviously a very large piece of work that will take some time to complete. We have identified the following steps that will be needed.
+
+* Prepare `frontend` to be the host
+    * convert frontend tests to RSpec
+    * modernise the tests
+    * audit `frontend` and make changes to the frontend directory structure to match “new layout”
+    * remove redundant code from frontend (non-blocking)
+    * remove publishing_api tasks from frontend (non-blocking)
+* Merge `feedback` into `frontend`
+    * use as a test case and use that to inform our next steps
+* Write a roadmap for app consolidation
+    * will be handled separately
+    * might be simply the order in which applications will be merged into the host
+    * or could be some apps merged together before merging into the host
+    * or some routes merging ahead of others
+* Consider and work on expected issues (see below)
+* Work to merge remaining apps
+
+### Steps as part of a merge
```

Comment:
You've flagged that the apps you're proposing to merge have inconsistent approaches to testing. And merging codebases as proposed here is one of the times we'd really benefit from having comprehensive test coverage. I wonder if there's a case for rewriting a candidate's tests in RSpec and ensuring good coverage prior to any merge in to `frontend`?

Reply:
You may be interested in this: https://github.com/alphagov/frontend/pull/4090 (the PR that rewrites Frontend's test suite, since it's currently minitest). Of the candidate apps to merge in, all of them already have RSpec as their teset runner with the exceptions of government-frontend (the biggie) and smart-answers (which as other people have noted, probably isn't in scope for this realistically).